codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/ServiceGenerator.java

```diff
@@ -0,0 +1,168 @@
+    private final GoWriter writer;
+    private final ServiceShape service;
+    private final List<GoIntegration> integrations;
+    private final List<RuntimeClientPlugin> runtimePlugins;
+    private final ApplicationProtocol applicationProtocol;
+
+    ServiceGenerator(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoWriter writer,
+            ServiceShape service,
+            List<GoIntegration> integrations,
+            List<RuntimeClientPlugin> runtimePlugins,
+            ApplicationProtocol applicationProtocol
+    ) {
+        this.settings = settings;
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.service = service;
+        this.integrations = integrations;
+        this.runtimePlugins = runtimePlugins;
+        this.applicationProtocol = applicationProtocol;
+    }
+
+    @Override
+    public void run() {
+        Symbol serviceSymbol = symbolProvider.toSymbol(service);
+        writer.writeShapeDocs(service);
+        writer.openBlock("type $T struct {", "}", serviceSymbol, () -> {
+            writer.write("options $L", CONFIG_NAME);
+        });
+
+        generateConstructor(serviceSymbol);
+
+        String clientId = CodegenUtils.getDefaultPackageImportName(settings.getModuleName());
+        String clientTitle = service.getTrait(TitleTrait.class).map(StringTrait::getValue).orElse(clientId);
+
+        writer.writeDocs("ServiceID returns the name of the identifier for the service API.");
+        writer.write("func (c $P) ServiceID() string { return $S }", serviceSymbol, clientId);
+
+        writer.writeDocs("ServiceName returns the full service title.");
+        writer.write("func (c $P) ServiceName() string { return $S }", serviceSymbol, clientTitle);
+
+        generateConfig();
+
+        Symbol stackSymbol = SymbolUtils.createPointableSymbolBuilder("Stack", GoDependency.SMITHY_MIDDLEWARE).build();
+        writer.write("type $L func($P) error", API_OPTIONS_FUNC_NAME, stackSymbol);
+    }
+
+    private void generateConstructor(Symbol serviceSymbol) {
+        writer.writeDocs(() -> writer.write(
+                    "New returns an initialized $T based on the functional options.\n"
+                    + "Provide additional functional options to further configure the behavior\n"
+                    + "of the client, such as changing the client's endpoint or adding custom\n"
+                    + "middleware behavior.", serviceSymbol
+        ));
+        writer.openBlock("func New(options $L) ($P, error) {", "}",
+                CONFIG_NAME, serviceSymbol, () -> {
+
+            writer.write("options = options.Copy()").write("");
+
+            // Run any config initialization functions registered by runtime plugins.
+            for (RuntimeClientPlugin runtimeClientPlugin : runtimePlugins) {
+                if (!runtimeClientPlugin.matchesService(model, service)
```

Comment:
You'll want to filter this set once in the constructor to pre-compute it when you need other aspects of the `RuntimeClientPlugin` class in here. Fine to fix up when you add that next integration.